### PR TITLE
Extracted union code from IndexInfo to IndexInfoUnion. Fixed a bug wh…

### PR DESCRIPTION
**Pull request by apmoriarty**
…ere delayed nodes would continue to be added under certain circumstances.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+ * The union of IndexInfo(A) and IndexInfo(B) produce the following,
+ * <p>
+ * <code>
+ * IndexInfo(A OR B)
+ * doc 1 = {A}
+ * doc 2 = {A, B}
+ * doc 3 = {B}
+ * </code>
+ */
+public class IndexInfoUnion {
```
Comment (jwomeara) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java]: Hmmmm, why did this need to be broken out into a separate class?  This makes tracking your changes a bit harder than it would be otherwise.  Plus, if we ever decided to extend IndexInfo, we wouldn't be able to extend these methods.

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java]: The original implementation was hard to follow and read like a utility method (taking another IndexInfo, returning a brand new one). I broke the union operation out for the refactor and decided to leave it out, a la the Single Responsibility Principle.

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java [hunk 2]
```diff
@@ -75,10 +71,24 @@ public IndexInfo(Iterable<?> ids) {
         this.count = this.uids.size();
     }
     
+    /**
+     * Helper method that determines if the IndexInfo contains only document ranges.
```
[the remark this answers is not in the document]
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java]: Got it

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static datawave.query.jexl.JexlASTHelper.nodeToKey;
+
+/**
+ * This class generates the union of two {@link IndexInfo} objects. It handles all combinations of document and infinite ranges and will add in any nodes that
+ * were delayed. These are nodes that the DefaultQueryPlanner determined were not appropriate for the index lookup, thus delaying them for later evaluation.
```
[the remark this answers is not in the document]
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java]: I'll move this to the DQP.

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        
+        // 1. Unwrap the top level Jexl node for the IndexInfo objects and merge them together.
+        mergedNodes = addJexlNodesToMap(mergedNodes, left.myNode);
+        mergedNodes = addJexlNodesToMap(mergedNodes, right.myNode);
+        
+        // 2. Add delayed nodes to the merged nodes
+        mergedNodes = addJexlNodesToMap(mergedNodes, delayedNodes);
+        
+        // 3. Generate the inverted multimap of uids to jexl nodes.
+        Set<IndexMatch> matches = mergeIndexMatches(left.uids, right.uids, delayedNodes);
```
[the remark this answers is not in the document]
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java]: Yes, I'll add a comment explaining why.

File: warehouse/query-core/src/test/java/datawave/query/index/lookup/IndexInfoTest.java [hunk 5]
```diff
@@ -276,28 +273,6 @@ public void testUnion_BothTermsHaveDocIds() {
         assertEquals(expectedDocs, merged.uids());
     }
     
-    /**
-     * Union of query terms when one term is a delayed predicate
-     */
-    @Test
-    public void testUnion_OneTermIsDelayedPredicate() {
```
[the remark this answers is not in the document]
Comment (apmoriarty) [reply warehouse/query-core/src/test/java/datawave/query/index/lookup/IndexInfoTest.java]: I talked with @ivakegg and we determined that this was testing for something that isn't possible.

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+     * @param left
+     * @param right
+     * @param delayedNodes
+     * @return
+     */
+    private static Set<IndexMatch> mergeIndexMatches(ImmutableSortedSet<IndexMatch> left, ImmutableSortedSet<IndexMatch> right, List<JexlNode> delayedNodes) {
+        HashMultimap<String,JexlNode> invertedNodeMap = HashMultimap.create();
+        HashMultimap<String,String> invertedNodeStringMap = HashMultimap.create();
+        for (IndexMatch match : Iterables.concat(left, right)) {
+            if (null != match.myNodes && !match.myNodes.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java]: I'll add a check to see if we have a null top-level jexl node.

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java (hunk 1 above)
Comment (alerman) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java]: I tend to agree with @apmoriarty on this one, but I feel like the naming feels funny to me. Not sure I like IndexInfoUnion as this class does not represent an IndexInfoUnion, it merely contains the method to generate one.

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+            merged.uids = ImmutableSortedSet.copyOf(matches);
+        }
+        
+        // 6. Handle uid count
+        if (left.isInfinite() || right.isInfinite()) {
+            merged.count = -1;
+        } else {
+            merged.count = merged.uids.size();
+        }
+
```
Comment (alerman) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java]: Could these be combined since the condition is the same, or is there something else I am missing?

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfoUnion.java]: They could be combined. I'll do that while resolving some rebase conflicts.
